src/irr/asset/CGraphicsPipelineLoaderMTL.cpp

```diff
@@ -12,61 +31,74 @@ void main()
     mat3 normalMat = irr_glsl_SBasicViewParameters_GetNormalMat(CamData.params);
     Normal = normalMat*normalize(vNormal);
 }
+#endif //_IRR_VERT_MAIN_DEFINED_
+)";
+    constexpr const char* FRAG_SHADER_NO_UV_INPUTS = 
+R"(
+#ifndef _IRR_FRAG_INPUTS_DEFINED_
+#define _IRR_FRAG_INPUTS_DEFINED_
+        layout(location = 0) in vec3 LocalPos;
+        layout(location = 1) in vec3 ViewPos;
+        layout(location = 2) in vec3 Normal;
+#endif //_IRR_FRAG_INPUTS_DEFINED_
 )";
     constexpr const char* FRAG_SHADER_NO_UV_BSDF_COS_EVAL =
-R"(#ifndef _IRR_BSDF_COS_EVAL_DEFINED_
+R"(
+#ifndef _IRR_BSDF_COS_EVAL_DEFINED_
+#ifndef _IRR_BSDF_COS_EVAL_DEFINED_
 #define _IRR_BSDF_COS_EVAL_DEFINED_
 
 // Spectrum can be exchanged to a float for monochrome
 #define Spectrum vec3
-#define Ia 0.1
 
 //! This is the function that evaluates the BSDF for specific view and observer direction
 // params can be either BSDFIsotropicParams or BSDFAnisotropicParams 
-Spectrum irr_glsl_bsdf_cos_eval(in irr_glsl_BSDFIsotropicParams params)
+Spectrum irr_bsdf_cos_eval(in irr_glsl_BSDFIsotropicParams params)
 {
     vec3 Kd = PC.params.Kd;
 
     vec3 color = vec3(0.0);
-    if ((PC.params.extra&ILLUM_MODEL_MASK) > 0)
-    {
-        vec3 Ka = PC.params.Ka;
-        vec3 Ks = PC.params.Ks;
-        float Ns = PC.params.Ns;
+    vec3 Ks = PC.params.Ks;
+    float Ns = PC.params.Ns;
 
-        vec3 diff = Kd*params.NdotL;
-        switch (PC.params.extra&ILLUM_MODEL_MASK)
-        {
-        case 1:
-            color = Ia*Ka + diff;
-            break;
-        case 2:
-        case 3://2 + IBL
-        case 5://basically same as 3
-        case 8://basically same as 5
-        {
-            float fr = Fresnel_dielectric(PC.params.Ni, params.NdotL);
-            vec3 spec = Ks*pow(params.NdotH, Ns);
-            color += Ia*Ka + mix(diff, spec, fr);
-        }
-            break;
-        case 4:
-        case 6:
-        case 7:
-        case 9://basically same as 4
-        {
-            float fr = Fresnel_dielectric(PC.params.Ni, params.NdotL);
-            vec3 spec = Ks*pow(params.NdotH, Ns);
-            color = fr*(Ia*Ka+spec);
-        }
-            break;
-        default:
-            break;
-        }
+    vec3 diff = irr_glsl_lambert() * Kd * (1.0-Fresnel_dielectric(PC.params.Ni,params.NdotL)) * (1.0-Fresnel_dielectric(PC.params.Ni,params.NdotV));
+    diff *= irr_glsl_diffuseFresnelCorrectionFactor(vec3(PC.params.Ni), vec3(PC.params.Ni*PC.params.Ni));
+    switch (PC.params.extra&ILLUM_MODEL_MASK)
+    {
+    case 0:
+        color = Kd;
+        break;
+    case 1:
+        color = diff * params.NdotL;
+        break;
+    case 2:
+    case 3://2 + IBL
+    case 5://basically same as 3
+    case 8://basically same as 5
+    {
+        float fr = Fresnel_dielectric(PC.params.Ni, params.VdotH);
+        //TODO normalize specular
```

Comment:
and provide from a bsdf/brdf built-in header as a cook-torrance-like function (so fresnel goes inside)

Reply:
do you mean another function that looks like
```
float f(...)
{
   return normalized_blinn_phong(...)*fresnel(...);
}
//+ another one for fresnel from complex ior
```
?

Reply:
pretty much.